dali/pipeline/data/tensor_vector.cc

```diff
@@ -842,118 +1019,145 @@ Tensor<Backend> TensorVector<Backend>::AsTensor() {
 
 
 template <typename Backend>
-void TensorVector<Backend>::UpdateViews() {
-  // Return if we do not have a valid allocation
-  if (!IsValidType(tl_->type())) return;
-  // we need to be able to share empty view as well so don't check if tl_ has any data
-  type_ = tl_->type_info();
-  sample_dim_ = tl_->shape().sample_dim();
-
-  assert(curr_num_tensors_ == tl_->num_samples());
-
-  for (int i = 0; i < curr_num_tensors_; i++) {
-    update_view(i);
-  }
+void TensorVector<Backend>::ShareData(const shared_ptr<void> &ptr, size_t bytes, bool pinned,
+                                      const TensorListShape<> &shape, DALIDataType type,
+                                      int device_id, AccessOrder order,
+                                      const TensorLayout &layout) {
+  contiguous_buffer_.set_backing_allocation(ptr, bytes, pinned, type, shape.num_elements(),
+                                            device_id, order);
+  buffer_bkp_.reset();
+  tensors_.clear();
+  tensors_.resize(shape.num_samples());
+
+  state_.Update(BatchState::Contiguous);
+  curr_num_tensors_ = shape.num_samples();
+  type_ = TypeTable::GetTypeInfo(type);
+  sample_dim_ = shape.sample_dim();
+  shape_ = shape;
+  layout_ = layout;
+  pinned_ = pinned;
+  device_ = device_id;
+  order_ = order;
+  recreate_views();
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::resize_tensors(int new_size) {
-  if (static_cast<size_t>(new_size) > tensors_.size()) {
+  // This doesn't update with the same order as the class members are listed
+  // We need to make sure everything is updated for the tensors that come back into scope
+  // and we start with the pinned and order properties as they might impact future allocations.
+  // next we make sure the type is consistent, and if so, introduce empty shape
+  shape_.resize(new_size);
+  if (new_size > curr_num_tensors_) {
     auto old_size = curr_num_tensors_;
     tensors_.resize(new_size);
     for (int i = old_size; i < new_size; i++) {
-      if (!tensors_[i]) {
-        tensors_[i] = std::make_shared<Tensor<Backend>>();
-        tensors_[i]->set_pinned(is_pinned());
-        tensors_[i]->set_order(order());
+      // TODO(klecki) same validation as when updating properties - or reset
+      if (!tensors_[i].has_data()) {
+        tensors_[i].set_pinned(is_pinned());
+      } else {
+        DALI_ENFORCE(tensors_[i].is_pinned() == is_pinned());
```

Comment:
Is this actually possible: to start with 10 samples, resize to 5 (or 0), change the pinnedness, resize to 10 and get the error that the sample that "has ceased to exist" has now different pinnedness and thus the resize failed?

Reply:
Tbh, this is probably a case that won't happen, as we rarely even touch the set_pinned, but I don't want to have wall of ifs for those weird cases - maybe it will be easier to just guard against it and worry about it when someone tries something "clever", like jumping between pinned and not pinned memory in the same variable-batch-sized buffer.

Reply:
Yeah, I guess it's reasonable not to prolong the refector. Still, looking at the TODO note I wonder what properites may, at this point, be obsolete and if they can be adjsuted. It seems super weird to fail the resize because some samples that from the outside "are not there". My vote would be for resetting any such sample if it turns out that it cannot be reused.

Reply:
Done.